File: frontend/src/types/chart.ts
export interface Planet {
  longitude: number;
  latitude: number;
  speed: number;
  is_retrograde: boolean;
  sign?: number;           // 1-12
  degree_in_sign?: number;
  house?: number;
  dignity?: string;
  dignity_list?: string[];
  speed_status?: string;
  is_combust?: boolean;
  planetary_war?: boolean;
  thai_name?: string;
  symbol?: string;
  lordships?: { house: number; name: string }[];
}

export interface Lagna {
  longitude: number;
  sign: number;
  cusps: number[];
}

export interface NakshatraInfo {
  index: number;
  name: string;
  pada: number;
}

export interface DashaPeriod {
  planet: string;
  start: string;
  end: string;
  is_current: boolean;
  antardashas?: DashaPeriod[];
}

export interface ChartData {
  julian_date: number;
  ayanamsa_name: string;
  ayanamsa_value: number;
  lagna: Lagna;
  planets: { [key: string]: Planet };
  houses: { [key: string]: number };
  house_lords: { [key: number]: { planet: string; sign: number; name: string } };
  d3_house_lords?: { [key: number]: { planet: string; sign: number; name: string } };
  d9_house_lords?: { [key: number]: { planet: string; sign: number; name: string } };
  yogas: any[];
  d3_yogas?: any[];
  d9_yogas?: any[];
  western_aspects: any[];
  d3_western_aspects?: any[];
  d9_western_aspects?: any[];
  vedic_aspects: any[];
  d3: { [key: string]: Planet };
  d9: { [key: string]: Planet };
  d3_lagna?: Lagna;
  d9_lagna?: Lagna;
  lunar_data: {
    moon_nakshatra: NakshatraInfo;
    lagna_nakshatra: NakshatraInfo;
    planet_nakshatras: { [key: string]: NakshatraInfo };
  };
  dasha_timeline: DashaPeriod[];
  advanced_analysis?: {
    inthaphas: { planet: string; sign: number; longitude: number };
    batchan: { planet: string; sign: number };
    is_dual_lord: boolean;
    summary_th: string;
  };
}

export interface BirthFormData {
  name?: string;
  year: number;
  month: number;
  day: number;
  hour: number;
  minute: number;
  lat: number;
  lon: number;
  timezone: string;
  ayanamsa_mode?: string;
  node_type?: "MEAN" | "TRUE";
  ketu_mode?: "vedic" | "thai";
  house_system?: string;
  aspect_orb?: number;
  planet_corrections?: { [key: string]: number };
}

export interface CompareResponse {
  person_a_chart: ChartData;
  person_b_chart: ChartData;
  synastry_aspects?: any[];
  compatibility_summary: {
    message: string;
    [key: string]: any;
  };
}

File: backend/models/chart.py
from pydantic import BaseModel, ConfigDict
from typing import Dict, Any, List

class BirthChart(BaseModel):
    """
    Immutable snapshot of a natal chart.
    Rule: Once created, this object must not be mutated.
    """
    model_config = ConfigDict(frozen=True)

    julian_date: float
    ayanamsa_name: str
    ayanamsa_value: float
    lagna: Dict[str, Any]
    planets: Dict[str, Any]
    houses: Dict[str, int] # House Number (str) → Sign (1-12)
    western_aspects: List[Dict[str, Any]]
    d3_western_aspects: List[Dict[str, Any]] | None = None
    d9_western_aspects: List[Dict[str, Any]] | None = None
    vedic_aspects: List[Dict[str, Any]]
    d3: Dict[str, Any]
    d9: Dict[str, Any]
    d3_lagna: Dict[str, Any] | None = None
    d9_lagna: Dict[str, Any] | None = None
    lunar_data: Dict[str, Any]
    dasha_timeline: List[Dict[str, Any]]
    house_lords: Dict[int, Any] | None = None
    d3_house_lords: Dict[int, Any] | None = None
    d9_house_lords: Dict[int, Any] | None = None
    yogas: List[Dict[str, Any]] | None = None
    d3_yogas: List[Dict[str, Any]] | None = None
    d9_yogas: List[Dict[str, Any]] | None = None



